org.eclipse.xtext.xtext.ui/src/org/eclipse/xtext/xtext/ui/editor/quickfix/XtextGrammarQuickfixProvider.java

```diff
@@ -134,24 +137,39 @@
 	@Inject
 	private ILineSeparatorInformation separatorInfo;
 	
+	@Inject
+	private XtextGrammarAccess grammarAccess;
+	
 	@Fix(XtextLinkingDiagnosticMessageProvider.UNRESOLVED_RULE)
 	public void fixUnresolvedRule(final Issue issue, IssueResolutionAcceptor acceptor) {
 		final String ruleName = issue.getData()[0];
-		acceptor.accept(issue, "Create rule '" + ruleName + "'", "Create rule '" + ruleName + "'", NULL_QUICKFIX_IMAGE,
-				new ISemanticModification() {
-					@Override
-					public void apply(final EObject element, IModificationContext context) throws BadLocationException {
-						AbstractRule abstractRule = EcoreUtil2.getContainerOfType(element, ParserRule.class);
-						ICompositeNode node = NodeModelUtils.getNode(abstractRule);
-						int offset = node.getEndOffset();
-						String nl = context.getXtextDocument().getLineDelimiter(0);
-						StringBuilder builder = new StringBuilder(nl+nl);
-						if (abstractRule instanceof TerminalRule)
-							builder.append("terminal ");
-						String newRule = builder.append(ruleName).append(":" + nl + "\t" + nl + ";").toString();
-						context.getXtextDocument().replace(offset, 0, newRule);
-					}
-				});
+		acceptor.accept(issue, // 
+				"Create rule '" + ruleName + "'", //
+				"Create rule '" + ruleName + "'", //
+				NULL_QUICKFIX_IMAGE, //
+				createNewRule(ruleName, false));
+		createLinkingIssueResolutions(issue, acceptor);
+	}
+
+	@Fix(XtextLinkingDiagnosticMessageProvider.UNRESOLVED_TERMINAL_RULE)
+	public void fixUnresolvedTerminalRule(final Issue issue, IssueResolutionAcceptor acceptor) {
+		final String ruleName = issue.getData()[0];
+		acceptor.accept(issue, //
+				"Create terminal rule '" + ruleName + "'", // 
+				"Create terminal rule '" + ruleName + "'", //
+				NULL_QUICKFIX_IMAGE, //
+				createNewRule(ruleName, false));
+		createLinkingIssueResolutions(issue, acceptor);
+	}
+	
+	@Fix(XtextLinkingDiagnosticMessageProvider.UNRESOLVED_TERMINAL_RULE)
+	public void fixUnresolvedTerminalFragmentRule(final Issue issue, IssueResolutionAcceptor acceptor) {
+		final String ruleName = issue.getData()[0];
+		acceptor.accept(issue, //
+				"Create terminal fragment rule '" + ruleName + "'", // 
```

Comment:
Please reword to `Create terminal fragment` instead of `Create terminal fragment rule`

Reply:
Done.